modules/core/src/controllers/controller.ts

```diff
@@ -61,71 +73,100 @@ type ControllerProps = {
-
-  makeViewport;
-  _eventStartBlocked;
-  _state;
-
-  _panMove;
-  invertPan;
-  dragMode: 'pan' | 'rotate' = 'rotate';
-
-  inertia: number = 0;
-
-  scrollZoom: boolean | {speed?: number; smooth?: boolean} = true;
-  dragPan: boolean = true;
-  dragRotate: boolean = true;
-  doubleClickZoom: boolean = true;
-  touchZoom: boolean = true;
-  touchRotate: boolean = false;
-  keyboard:
+  private _customEvents: string[] = [];
+  private _eventStartBlocked: any = null;
+  private _panMove: boolean = false;
+
+  protected invertPan: boolean = false;
+  protected dragMode: 'pan' | 'rotate' = 'rotate';
+  protected inertia: number = 0;
+  protected scrollZoom: boolean | {speed?: number; smooth?: boolean} = true;
+  protected dragPan: boolean = true;
+  protected dragRotate: boolean = true;
+  protected doubleClickZoom: boolean = true;
+  protected touchZoom: boolean = true;
+  protected touchRotate: boolean = false;
+  protected keyboard:
     | boolean
     | {
-        zoomSpeed: number; //  speed of zoom using +/- keys. Default 2.
-        moveSpeed: number; //  speed of movement using arrow keys, in pixels.
-        rotateSpeedX: number; //  speed of rotation using shift + left/right arrow keys, in degrees. Default 15.
-        rotateSpeedY: number; //  speed of rotation using shift + up/down arrow keys, in degrees. Default 10.
+        zoomSpeed?: number; //  speed of zoom using +/- keys. Default 2.
+        moveSpeed?: number; //  speed of movement using arrow keys, in pixels.
+        rotateSpeedX?: number; //  speed of rotation using shift + left/right arrow keys, in degrees. Default 15.
+        rotateSpeedY?: number; //  speed of rotation using shift + up/down arrow keys, in degrees. Default 10.
       } = true;
 
-  constructor(ControllerState, options: ControllerProps = {}) {
+  constructor(ControllerState: ConstructorOf<ControllerState>, props: ControllerProps & {
+    timeline: Timeline,
+    eventManager: EventManager;
+    makeViewport: (opts: Record<string, any>) => Viewport;
+    onViewStateChange: (params: ViewStateChangeParameters) => void;
+    onStateChange: (state: InteractionState) => void;
+  }) {
     this.ControllerState = ControllerState;
-    this.transitionManager = new TransitionManager(ControllerState, {
-      ...options,
+    this.transitionManager = new TransitionManager<ControllerState>(ControllerState, {
+      ...props,
       onViewStateChange: this._onTransition.bind(this),
       onStateChange: this._setInteractionState.bind(this)
     });
 
     const linearTransitionProps = this.linearTransitionProps;
     this._transition = linearTransitionProps && {
       ...LINEAR_TRANSITION_PROPS,
-      // @ts-expect-error
       transitionInterpolator: new LinearInterpolator({
         transitionProps: linearTransitionProps
       })
     };
 
     this.handleEvent = this.handleEvent.bind(this);
 
-    this.setProps(options);
+    this.props = props;
+    this.eventManager = props.eventManager;
```

Comment:
These would no longer be needed if props are `Required`. you can just call `this.props.onViewStateChange`

Reply:
`eventManager` / `onViewStateChange` etc are defined once in the constructor and  never again when `setProps` are called.

Reply:
Yes I see... though perhaps that wouldn't prevent the `Required<Props>` technique from being used if we filtered out (or throw an error for) any initialization-only fields in `setProps()`. 

Initialization only props are of course a bit special, but it is kind of weird to allow the `this.props` field to diverge from the actual truth.

FWIW, in luma.gl I did find that being able to count on `this.props` having required fields simplified writing type-safe code accessing those fields, sometimes it could be surprisingly annoying to have to deal with optional props.